xmtp_mls/src/utils/test/mod.rs

```diff
@@ -336,3 +338,56 @@ pub async fn wait_for_min_intents(conn: &DbConnection, n: usize) {
         published = conn.intents_published() as usize;
     }
 }
+
+#[cfg(any(test, feature = "test-utils"))]
+/// Checks if test mode is enabled.
+pub fn is_test_mode_upload_malformed_keypackage() -> bool {
+    env::var("TEST_MODE_UPLOAD_MALFORMED_KP").unwrap_or_else(|_| "false".to_string()) == "true"
+}
+
+#[cfg(any(test, feature = "test-utils"))]
+#[warn(dead_code)]
+/// Sets test mode and specifies malformed installations dynamically.
+/// If `enable` is `false`, it also clears `TEST_MODE_MALFORMED_INSTALLATIONS`.
+pub fn set_test_mode_upload_malformed_keypackage(
+    enable: bool,
+    installations: Option<Vec<Vec<u8>>>,
+) {
+    if enable {
+        env::set_var("TEST_MODE_UPLOAD_MALFORMED_KP", "true");
+
+        if let Some(installs) = installations {
+            let installations_str = installs
+                .iter()
+                .map(hex::encode)
+                .collect::<Vec<_>>()
+                .join(",");
+
+            env::set_var("TEST_MODE_MALFORMED_INSTALLATIONS", installations_str);
+        }
+    } else {
+        env::set_var("TEST_MODE_UPLOAD_MALFORMED_KP", "false");
+        env::remove_var("TEST_MODE_MALFORMED_INSTALLATIONS");
+    }
+}
+
+#[cfg(any(test, feature = "test-utils"))]
+/// Retrieves and decodes malformed installations from the environment variable.
+/// Returns an empty list if test mode is not enabled.
+pub fn get_test_mode_malformed_installations() -> Vec<Vec<u8>> {
+    if !is_test_mode_upload_malformed_keypackage() {
+        return Vec::new();
+    }
+
+    env::var("TEST_MODE_MALFORMED_INSTALLATIONS")
```

Comment:
how do these environment variables work? What am I expected to set here

Reply:
this is only for dynamic testing purposes, see here:
https://github.com/xmtp/libxmtp/pull/1606/files#diff-c29f56a38916c7410eff8091df1a2e43487ffe20646d96827e846e475f4608d3R2428

made this to manipulate the logic during the tests to simulate certain conditions.